File: 1124.Longest Well-Performing Interval.py
class Solution(object):
    def longestWPI(self, hours):
        """
        :type hours: List[int]
        :rtype: int
        """
        points = [1 if x > 8 else -1 for x in hours]
        seen = {0: -1}
        benefit = 0
        res = 0
        for i, point in enumerate(points):
            benefit += point
            seen.setdefault(benefit, i)
            if benefit > 0:
                res = max(res, i + 1)
            else:
                res = max(res, i - seen.get(benefit - 1, float('inf')))
        print seen
        return res

File: 1124.Longest Well-Performing Interval.java
class Solution {
    public int longestWPI(int[] hours) {
        int res = 0;
        int benefit = 0;
        List<Integer> seen = new ArrayList<>();
        for (int i = 0; i < hours.length; i++) {
            benefit += hours[i] > 8 ? 1 : -1;
            if (benefit < 0 && -benefit > seen.size()) {
                seen.add(i);
            }
            if (benefit > 0) {
                res = Math.max(res, i + 1);
            } else {
                res = Math.max(res, i - (seen.size() > -benefit ? seen.get(-benefit) : Integer.MIN_VALUE));
            }
        }
        return res;
    }
}
